nmmo/task/game_state.py

```diff
@@ -0,0 +1,94 @@
+from typing import Dict, List, Tuple
+from dataclasses import dataclass
+from copy import deepcopy
+
+import numpy as np
+import numpy_indexed as npi
+
+from nmmo.core.config import Config
+from nmmo.core.realm import Realm
+from nmmo.core.observation import Observation
+
+from nmmo.entity.entity import EntityState
+from nmmo.systems.item import ItemState
+from nmmo.core.tile import TileState
+
+
+@dataclass
+class GameState:
+  tick: int
+  config: Config
+  spawn_pos: Dict[int, Tuple[int, int]] # ent_id: (row, col)
+
+  alive_agents: List # of alive agents' ent_id
+  env_obs: Dict[int, Observation] # this should include only alive agents
+
+  entity_cols: Dict # attr_to_col
+  entity_data: np.ndarray # a copied, whole Entity ds table
+
+  item_cols: Dict # attr_to_col
+  item_data: np.ndarray # a copied, whole Item ds table
+
+  tile_cols: Dict
+
+  cache_result: Dict # cache for result of Predicate results
+
+  # add extra info that is not in the datastore (e.g., spawn pos)
+  # add helper functions below
+
+  def entity_or_none(self, ent_id):
+    flt_ent = self.entity_data[:,self.entity_cols['id']] == ent_id
+    if np.any(flt_ent):
+      return EntityState.parse_array(self.entity_data[flt_ent][0])
+
+    return None
+
+  def where_in_id(self, data_type, subject: List[int]):
+    if data_type == 'entity':
+      flt_idx = np.in1d(self.entity_data[:,self.entity_cols['id']], subject)
+      return self.entity_data[flt_idx]
+
+    if data_type == 'item':
+      flt_idx = np.in1d(self.item_data[:,self.item_cols['owner_id']], subject)
+      return self.item_data[flt_idx]
+
+    return None
+
+  def group_by(self, flt_data, grpby_col, sum_col=0):
+    # if sum_col = 0, this fn acts as COUNT, otherwise SUM
+    g = npi.group_by(flt_data[:,grpby_col])
+    result = {}
+    for k, v in zip(*g(flt_data[:,sum_col])):
+      if sum_col:
+        result[k] = sum(v)
+      else:
+        result[k] = len(v)
+    return result
+
+
+class GameStateGenerator:
+  def __init__(self, realm: Realm, config: Config):
+    self.config = deepcopy(config)
+    self.spawn_pos: Dict[int, Tuple[int, int]] = {}
+
+    for ent_id, ent in realm.players.items():
+      self.spawn_pos.update( {ent_id: ent.pos} )
+
+  def generate(self, realm: Realm, env_obs: Dict[int, Observation]) -> GameState:
+    # copy the datastore, by running astype
+    entity_all = EntityState.Query.table(realm.datastore).astype(np.int16)
+    ent_cols = deepcopy(EntityState.State.attr_name_to_col)
+
+    return GameState(
+      tick = deepcopy(realm.tick),
+      config = deepcopy(self.config),
```

Comment:
deep copy seems expensive. why can't we just provide a reference to the config?

Reply:
Removed deepcopy for the config, since it was deep-copied when the game state generator was initiated. Messing with it won't affect the game.

The other things that are deep-copied are: env_obs, spawn_pos, entity_data, and item_data (these numpy tables are copied via astype(np.int16) function) 
* Messing with env_obs can affect the next step, so it'd be good to protect it.
* Messing with spawn_pos can affect the `DistanceTraveled` predicate. I don't think it's high risk, but since it's small, so why not?

Reply:
Chatgpt taught me `@dataclass(frozen=True)`, which can keep everything read only without deep copy.

gs.cache_result is still available to write via `MutableMapping`. Cool.